File: frontend/src/lib/workflowConfig.ts
import api from './api'

export interface StageSettings {
  visible_fields?: string[] // Fields from submission that are visible
  email_notifications?: {
    enabled: boolean
    recipients?: ('user' | 'vendor' | 'next_approver')[]
    reminders?: number[] // Days before reminder (e.g., [1, 2])
  }
  layout_id?: string // Form layout ID for approver screen tabs
}

export interface WorkflowStep {
  step_number: number
  step_type: 'approval' | 'notification'  // Removed 'review' - everything is approval now
  step_name: string
  assigned_role?: string
  assigned_user_id?: string
  approver_group_id?: string
  required: boolean
  can_skip: boolean
  auto_assign: boolean
  conditions?: Record<string, any>
  is_first_step?: boolean
  assignment_rule?: {
    type: 'role' | 'user' | 'group' | 'round_robin'
    value?: string // user_id, role, or group_id
  }
  stage_settings?: StageSettings
}

export interface ApproverGroup {
  id: string
  name: string
  description?: string
  tenant_id: string
  member_ids: string[]
  created_at: string
  updated_at: string
}

export interface AssignmentRules {
  approver_selection: 'round_robin' | 'specific_user' | 'role_based' | 'group_based'
  specific_approver_id?: string
  approver_group_id?: string
  reviewer_auto_assign: boolean
  escalation_rules?: {
    timeout_hours?: number
    escalate_to?: string
  }
}

export interface WorkflowConditions {
  agent_types?: string[]
  risk_levels?: string[]
  categories?: string[]
  priority: number
}

export interface TriggerRules {
  sso_groups?: string[]
  departments?: string[]
  application_categories?: string[]
  agent_types?: string[]
  risk_levels?: string[]
  match_all?: boolean // true = all conditions must match, false = any condition matches
}

export interface WorkflowConfig {
  id: string
  tenant_id: string
  name: string
  description?: string
  workflow_engine: 'internal' | 'servicenow' | 'jira' | 'custom'
  integration_id?: string
  integration_name?: string
  workflow_steps?: WorkflowStep[]
  assignment_rules?: AssignmentRules
  conditions?: WorkflowConditions
  trigger_rules?: TriggerRules
  status: 'active' | 'inactive' | 'draft'
  is_default: boolean
  created_by?: string
  created_at: string
  updated_at: string
}

export interface WorkflowConfigCreate {
  name: string
  description?: string
  workflow_engine?: 'internal' | 'servicenow' | 'jira' | 'custom'
  integration_id?: string
  integration_config?: Record<string, any>
  workflow_steps?: WorkflowStep[]
  assignment_rules?: AssignmentRules
  conditions?: WorkflowConditions
  trigger_rules?: TriggerRules
  is_default?: boolean
  status?: 'active' | 'inactive' | 'draft'
}

export interface OnboardingRequest {
  id: string
  agent_id: string
  tenant_id: string
  requested_by: string
  request_type: 'onboarding' | 'renewal' | 'update'
  workflow_config_id?: string
  workflow_engine: string
  external_workflow_id?: string
  status: 'pending' | 'in_review' | 'approved' | 'rejected' | 'cancelled'
  assigned_to?: string
  assigned_to_email?: string  // Email of assigned user
  assigned_to_name?: string  // Name of assigned user
  current_step: number
  reviewed_by?: string
  reviewed_at?: string
  review_notes?: string
  approved_by?: string
  approved_by_name?: string  // Name of approver
  approved_by_email?: string  // Email of approver
  approved_at?: string
  approval_notes?: string
  rejected_by?: string
  rejected_by_name?: string  // Name of rejector
  rejected_by_email?: string  // Email of rejector
  rejected_at?: string
  rejection_reason?: string
  request_number?: string  // Human-readable request number (AI-1, AI-2, etc.)
  created_at: string
  updated_at: string
}

export const workflowConfigApi = {
  list: async (): Promise<WorkflowConfig[]> => {
    const response = await api.get('/workflow-config')
    return response.data
  },

  get: async (id: string): Promise<WorkflowConfig> => {
    const response = await api.get(`/workflow-config/${id}`)
    return response.data
  },

  create: async (data: WorkflowConfigCreate): Promise<WorkflowConfig> => {
    const response = await api.post('/workflow-config', data)
    return response.data
  },

  update: async (id: string, data: Partial<WorkflowConfigCreate>): Promise<WorkflowConfig> => {
    const response = await api.patch(`/workflow-config/${id}`, data)
    return response.data
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/workflow-config/${id}`)
  },

  setFirstStep: async (id: string, stepNumber: number): Promise<WorkflowConfig> => {
    const response = await api.post(`/workflow-config/${id}/set-first-step?step_number=${stepNumber}`)
    return response.data
  },

  reorderSteps: async (id: string, stepOrder: number[]): Promise<WorkflowConfig> => {
    const response = await api.post(`/workflow-config/${id}/reorder-steps`, stepOrder)
    return response.data
  },

  listOnboardingRequests: async (status?: string): Promise<OnboardingRequest[]> => {
    const response = await api.get('/workflow-config/onboarding-requests', {
      params: status ? { status } : {}
    })
    return response.data
  },

  approveOnboardingRequest: async (requestId: string, notes?: string): Promise<OnboardingRequest> => {
    const response = await api.post(`/workflow-config/onboarding-requests/${requestId}/approve`, notes ? { notes } : null)
    return response.data
  },

  rejectOnboardingRequest: async (requestId: string, reason: string): Promise<OnboardingRequest> => {
    const response = await api.post(`/workflow-config/onboarding-requests/${requestId}/reject`, { reason })
    return response.data
  },

  getOnboardingRequestByAgent: async (agentId: string): Promise<OnboardingRequest | null> => {
    try {
      const response = await api.get(`/workflow-config/onboarding-requests/agent/${agentId}`)
      return response.data
    } catch (error: any) {
      // Handle 204 No Content or 404 - agent exists but no workflow request yet
      if (error?.response?.status === 204 || error?.response?.status === 404) {
        return null
      }
      throw error
    }
  },

  // Approver Groups
  listApproverGroups: async (): Promise<ApproverGroup[]> => {
    const response = await api.get('/workflow-config/approver-groups')
    return response.data
  },

  getApproverGroup: async (id: string): Promise<ApproverGroup> => {
    const response = await api.get(`/workflow-config/approver-groups/${id}`)
    return response.data
  },

  createApproverGroup: async (data: Omit<ApproverGroup, 'id' | 'created_at' | 'updated_at'>): Promise<ApproverGroup> => {
    const response = await api.post('/workflow-config/approver-groups', data)
    return response.data
  },

  updateApproverGroup: async (id: string, data: Partial<Omit<ApproverGroup, 'id' | 'created_at' | 'updated_at'>>): Promise<ApproverGroup> => {
    const response = await api.patch(`/workflow-config/approver-groups/${id}`, data)
    return response.data
  },

  deleteApproverGroup: async (id: string): Promise<void> => {
    await api.delete(`/workflow-config/approver-groups/${id}`)
  },

  healthCheck: async (): Promise<{
    tables_exist: Record<string, boolean>
    tenants: Array<{
      id: string
      name: string
      slug: string
      has_default_workflow: boolean
      has_any_workflow: boolean
      has_active_workflow: boolean
      workflow_count: number
      default_workflow?: {
        id: string
        name: string
        status: string
        steps_count: number
        engine: string
      }
    }>
    summary: {
      all_tables_exist: boolean
      total_tenants: number
      tenants_with_workflows: number
      tenants_without_workflows: number
      total_workflows: number
      total_onboarding_requests: number
    }
  }> => {
    const response = await api.get('/workflow-config/health-check')
    return response.data
  },
}

File: backend/app/models/workflow_config.py
"""
Workflow configuration models for tenant-specific workflow management
"""
from sqlalchemy import Column, String, DateTime, Text, ForeignKey, Boolean, JSON, Enum, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime
from app.core.database import Base
import enum


class WorkflowEngineType(str, enum.Enum):
    """Workflow engine types"""
    INTERNAL = "internal"  # Use platform's built-in workflow
    SERVICENOW = "servicenow"  # Integrate with ServiceNow
    JIRA = "jira"  # Integrate with Jira
    CUSTOM = "custom"  # Custom workflow engine via API


class WorkflowConfigStatus(str, enum.Enum):
    """Workflow configuration status"""
    ACTIVE = "active"
    INACTIVE = "inactive"
    DRAFT = "draft"


class WorkflowConfiguration(Base):
    """Workflow configuration for tenant"""
    __tablename__ = "workflow_configurations"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    
    # Workflow details
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    
    # Workflow engine selection
    workflow_engine = Column(String(50), nullable=False, default=WorkflowEngineType.INTERNAL.value)
    
    # Integration configuration (for external engines)
    # Note: Foreign key is optional - integration may not exist yet
    integration_id = Column(UUID(as_uuid=True), nullable=True)  # ForeignKey("integrations.id") - made optional to avoid import issues
    integration_config = Column(JSON, nullable=True)  # Engine-specific configuration
    
    # Workflow definition (for internal engine)
    workflow_steps = Column(JSON, nullable=True)  # Array of workflow steps
    # Example structure:
    # [
    #   {
    #     "step_number": 1,
    #     "step_type": "review",
    #     "step_name": "Security Review",
    #     "assigned_role": "security_reviewer",
    #     "required": true,
    #     "can_skip": false,
    #     "auto_assign": true,
    #     "stage_settings": {
    #       "visible_fields": ["name", "type", "llm_vendor", "data_sharing_scope", "capabilities"],
    #       "email_notifications": {
    #         "enabled": true,
    #         "recipients": ["user", "vendor", "next_approver"],
    #         "reminders": [1, 2]  # Days before reminder
    #       }
    #     }
    #   }
    # ]
    
    # Assignment rules
    assignment_rules = Column(JSON, nullable=True)  # Rules for assigning approvers/reviewers
    # Example structure:
    # {
    #   "approver_selection": "round_robin",  # round_robin, specific_user, role_based
    #   "specific_approver_id": null,
    #   "reviewer_auto_assign": true,
    #   "escalation_rules": {
    #     "timeout_hours": 48,
    #     "escalate_to": "tenant_admin"
    #   }
    # }
    
    # Conditions for applying this workflow
    conditions = Column(JSON, nullable=True)  # When to use this workflow
    # Example structure:
    # {
    #   "agent_types": ["ai", "automation"],
    #   "risk_levels": ["high", "critical"],
    #   "categories": ["security", "compliance"],
    #   "priority": 1  # Lower number = higher priority
    # }
    
    # Trigger rules for workflow routing
    trigger_rules = Column(JSON, nullable=True)  # Rules that trigger this workflow
    # Example structure:
    # {
    #   "sso_groups": ["security-team", "compliance-team"],
    #   "departments": ["IT", "Security", "Compliance"],
    #   "application_categories": ["security", "compliance", "data-processing"],
    #   "agent_types": ["AI_AGENT", "BOT"],
    #   "risk_levels": ["high", "critical"],
    #   "match_all": false  # true = all conditions must match, false = any condition matches
    # }
    
    # Status
    status = Column(String(50), nullable=False, default=WorkflowConfigStatus.DRAFT.value)
    is_default = Column(Boolean, default=False)  # Default workflow for tenant
    
    # Metadata
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # tenant = relationship("Tenant", back_populates="workflow_configurations")
    # integration = relationship("Integration", foreign_keys=[integration_id])


class OnboardingRequest(Base):
    """Onboarding request for agents (can be approved/reviewed/denied)"""
    __tablename__ = "onboarding_requests"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_id = Column(UUID(as_uuid=True), ForeignKey("agents.id"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    
    # Human-readable request number (e.g., AI-1, AI-2, AI-3)
    request_number = Column(String(50), nullable=True, unique=True, index=True)
    
    # Request details
    requested_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    request_type = Column(String(50), nullable=False, default="onboarding")  # onboarding, renewal, update
    
    # Invitation tracking
    invitation_id = Column(UUID(as_uuid=True), ForeignKey("vendor_invitations.id"), nullable=True, index=True)
    business_contact_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True)  # Person who invited the vendor
    
    # Workflow tracking
    workflow_config_id = Column(UUID(as_uuid=True), ForeignKey("workflow_configurations.id"), nullable=True)
    workflow_engine = Column(String(50), nullable=False, default=WorkflowEngineType.INTERNAL.value)
    external_workflow_id = Column(String(255), nullable=True)  # ID in external system (ServiceNow/Jira)
    
    # Status
    status = Column(String(50), nullable=False, default="pending")  # pending, in_review, approved, rejected, cancelled
    
    # Current assignee
    assigned_to = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True)
    current_step = Column(Integer, default=0)
    
    # Decision details
    reviewed_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    reviewed_at = Column(DateTime, nullable=True)
    review_notes = Column(Text, nullable=True)
    
    approved_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    approved_at = Column(DateTime, nullable=True)
    approval_notes = Column(Text, nullable=True)
    
    rejected_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    rejected_at = Column(DateTime, nullable=True)
    rejection_reason = Column(Text, nullable=True)
    
    # Metadata
    request_metadata = Column(JSON, nullable=True)  # Additional request data
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # agent = relationship("Agent", foreign_keys=[agent_id])
    # requester = relationship("User", foreign_keys=[requested_by])
    # workflow_config = relationship("WorkflowConfiguration", foreign_keys=[workflow_config_id])


class ApproverGroup(Base):
    """Approver groups for workflow assignment"""
    __tablename__ = "approver_groups"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    member_ids = Column(JSON, nullable=False)  # Array of user IDs
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # tenant = relationship("Tenant", back_populates="approver_groups")

